Add in-game updater for linux platforms

Diyou opened the pull request:
- Addresses https://github.com/ppy/osu/discussions/14653
- Supersedes #19109
- [x] Reworked AppImageUpdateTool and moved it to osu.Desktop/Linux
- [x] Moved shared code into the base class UpateManager

~~Renamed UpdateCompleteNotification that used to be a private class of SquirrelUpdateManager to ProgressCompleteNotification in the base class UpdateManager ( you might find a more suitable name since its derived from ProgressCompletionNotification.~~

TODOs:

- [x] Implement logging
- [ ] Addressing [recommended-user-experience](https://docs.appimage.org/packaging-guide/optional/updates.html#recommended-user-experience) ?

Diyou committed:
Applied CodeAnalysis

Diyou committed:
Update OsuGameDesktop.cs

Diyou committed:
Update

Diyou committed:
Reworked the linux updater

+ Reworked AppImageUpdateTool
+ Moved it into osu.Desktop/Linux
+ Moved sharable code into base UpdateManager

Diyou committed:
Reviewed ApplyUpdate

frenzibyte commented:
There's no need to open a whole new PR with changes addressed from the previous one? But seeing that this one is PR'd off another branch I guess it's for the better.

Diyou commented:
> But seeing that this one is PR'd off a topic branch rather than `master` I guess it's for the better.
That was the rationale at least. Hope this makes it more maintainable.

Diyou committed:
Replaced AppImageUpdateTool in favor of AppImageUpdateNET

Diyou commented:
I created a [Package](https://github.com/Diyou/AppImageUpdateNET) which wraps the native c++ code for c# via SWIG and added a class that implements those functions for easier use.

This adds better logging capabilities and removes the need to parse CLI output.

The package delivers a runtime library which needs to be copied to the binary folder when shipped (i assume this is an automated process)

peppy commented on osu.Desktop/Linux/.gitignore:
```diff
@@ -0,0 +1 @@
+!*.nupkg
```
If we're adding this, it should be in the global `.gitignore`. Not sure it's required (if this is bothering you then you can set an ignore in your home directory gitignore file.

Diyou replied:
I added this along the package for this branch. Both should be removed eventually.

Diyou replied:
I removed the gitignore since it does not have any purpose.

peppy reviewed (commented):
Haven't read through the changes, just thought I'd point this one out.

Diyou commented:
Something I wanted to note is that since the initial PR I could not manage to update to any new release with any external tool or this code due to missing public keys.
2022.731.1 seems to be signed but lacks the detached signature which results in a failed validation.
2022.709.1 was the last release I can conduct a successful update.

In code the update information can be changed like:
```csharp
updateManager ??= new UpdateManager(new UpdaterOptions
{
    AppPath = appPath
})
{
    Logger = new AppImageLogger(),
    RawUpdateInformation = "gh-releases-zsync|ppy|osu|2022.709.1|osu.AppImage.zsync"
};
```
I assume the process to generate the release image got changed?

Diyou committed:
Delete .gitignore

Diyou committed:
Update osu.Desktop.csproj

Diyou committed:
Update AppImageUpdateManager.cs

Changed the way the log storage gets accessed.

Diyou committed:
Moved shareable code into AbstractUpdateManager

+ restored original UpdateManager class and UpdateCompleteNotification to prevent future conflicts

Diyou committed:
Merge branch 'master' into linux_updater
